### DISCUSSION TITLE
How to deal with Cloudflare captcha on Login form?

### DISCUSSION THREAD

**Discussion by GeorgeXCV** [cypress-io/cypress#26945] [Questions and Help]
We recently added this Cloudflare verification to the login form. We login via API for 99% of our tests but still want a few tests that use the actual login form so we know if it's broken. I updated the test to click the captcha but it just get stuck in a loop. Has anyone dealt with this before?

<img width="287" alt="Screenshot 2023-06-06 at 15 01 39" src="https://github.com/cypress-io/cypress/assets/13811547/6844a883-5e7b-4f15-a794-094ca76b249c">
<img width="219" alt="Screenshot 2023-06-06 at 11 33 13" src="https://github.com/cypress-io/cypress/assets/13811547/a7fd06bf-fbfc-47d7-a0b6-857c21c0f14d">

**Comment by portyu9** [comment]
Hi @GeorgeXCV 

You should **not try to solve the real Cloudflare challenge from Cypress**. Cloudflare explicitly detects automated browsers such as Cypress, so a live challenge can loop or behave nondeterministically by design.

For Turnstile, Cloudflare provides test keys specifically for automated testing. For example, the visible "always passes" sitekey is:

```text
1x00000000000000000000AA
```

There is also an "always fails" test key, which is useful for testing the error path.

The recommended setup is therefore:

1. Use Cloudflare's test sitekey/secret in your test or staging environment.
2. Exercise the real login form with Cypress using that deterministic challenge.
3. Keep your programmatic/API login for the other 99% of the test suite, as you are already doing.
4. Optionally have a test using the failing key to verify your CAPTCHA error handling.

If this is a different Cloudflare managed challenge rather than Turnstile, the same principle applies: configure the test environment so the anti-bot challenge is bypassed/disabled for your test traffic rather than attempting to automate the production challenge.

Options such as `chromeWebSecurity: false` will not make a Cloudflare bot challenge reliable because the challenge is Cloudflare behavior, not Cypress's same-origin security.

Please mark this answer as accepted if it helped, thank you.
